File: Data Structures & Algorithms/eating-bananas/submission-0.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int l = 1, r = std::ranges::max(piles);
        int res = r;
        while(l <= r) {
            int m = l + ((r - l)/2);
            int t = consumeTime(piles, m);
            if(t > h)
                l = m + 1;
            else{
                res = m;
                r = m - 1;
            }
        }
        return res;
    }

private:
    int consumeTime(vector<int>& piles, int r) {
        int res = 0;
        for(int p: piles) {
            res += ((p + r - 1)/r);
        }
        return res;
    }
};

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-0.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        dp = {}
        def dfs(i, j):
            if (i, j) in dp:
                return dp[(i, j)]
            res = 1

            for di, dj in [[1, 0], [-1, 0], [0, 1], [0, -1]]:
                newI, newJ = i + di, j + dj
                if (
                    newI in range(len(matrix))
                    and newJ in range(len(matrix[0]))
                    and matrix[newI][newJ] < matrix[i][j]
                ):
                    res = max(res, 1 + dfs(newI, newJ))
            
            dp[(i, j)] = res
            return res
        
        res = 1
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                res = max(res, dfs(i, j))
        
        return res